File: src/main/java/com/frank/entity/book/BookShelf.java
package com.frank.entity.book;

import java.util.ArrayList;
import java.util.List;

public class BookShelf {

    final private List<Book> bookInShelf;

    final private int bookPlaceCount;

    private BookGenre genre;

    public BookShelf(int bookPlaceCount) {
        this.bookPlaceCount = bookPlaceCount;
        this.bookInShelf = new ArrayList<>(bookPlaceCount);
    }

    public BookShelf(int bookPlaceCount, BookGenre genre) {
        this.bookPlaceCount = bookPlaceCount;
        this.bookInShelf = new ArrayList<>(bookPlaceCount);
        this.genre = genre;
    }

    public void place(BookCollection bookCollection) {
        if (bookPlaceCount - bookInShelf.size() < 0) {
            throw new RuntimeException("No space in shelf " + this + " to store the collection " + bookCollection);
        }
        bookCollection.stream().forEach( book -> {
            bookInShelf.add(book);
            book.setGenre(this.genre);
            book.setShelf(this);
        });
    }

    public void remove(BookCollection bookCollection) {
        bookCollection.stream().forEach((book) -> {
            this.bookInShelf.remove(book.getId());
            book.setShelf(null);
        });
    }

    public boolean contains(Book book) {
        return bookInShelf.contains(book);
    }

    public boolean contains(BookCollection bookCollection) {
        return bookCollection.stream().allMatch((book) -> bookInShelf.contains(book) );
    }

}

File: src/main/kotlin/com/frank/entity/book/BookShelf.kt
package com.frank.entity.book

import java.util.ArrayList

class BookShelf(private val bookPlaceCount: Int, private val genre: BookGenre) {

    private val bookInShelf: MutableList<Book>

    init {
        this.bookInShelf = ArrayList(bookPlaceCount)
    }

    fun place(bookCollection: BookCollection) {
        if (bookPlaceCount - bookInShelf.size < 0) {
            throw RuntimeException("No space in shelf " + this + " to store the collection " + bookCollection)
        }
        bookCollection.asList().forEach { book ->
            bookInShelf.add(book)
            book.genre = this.genre
            book.shelf = this
        }
    }

    fun remove(bookCollection: BookCollection) {
        bookCollection.asList().forEach { book ->
            this.bookInShelf.remove(book.id)
            book.shelf = null
        }
    }

    operator fun contains(book: Book): Boolean {
        return bookInShelf.contains(book)
    }

    operator fun contains(bookCollection: BookCollection): Boolean {
        return bookCollection.asList().stream().allMatch { book -> bookInShelf.contains(book) }
    }

}
